### ISSUE TITLE
Global AssertionOptions are not thread-safe

### ISSUE THREAD

**Issue by philippdolder** [fluentassertions/fluentassertions#1323]
### Description
As discussed in #808 I'm opening a separate issue.

Global AssertionOptions appear to not be thread-safe.

Assume the following scenario.

Test 1 is using `Should().BeEquivalentTo()` and the current execution is at `CloneDefaults`
Test 2 is executing `AssertionOptions.AssertEquivalencyUsing()` at the same time on a different thread. This modifies the collection the `CloneDefaults` is currently iterating through. Which is of course not allowed.

### Complete minimal example reproducing the issue

Since this is a multi-threading issue it's a bit difficult to reproduce. The more tests there are run in parallel, the higher the chance it will occur. But basically this is an outline of what happens in various tests that eventually lead to the exception. I'm trying to get a failure with only these 2 tests using Riders `Run tests until failure` feature

```csharp
namespace FluentAssertionsRepro
{
    using FluentAssertions;
    using Xunit;

    public class Test1
    {
        [Fact]
        public void Test()
        {
            const string Text = "Some text";
            var container = new Container(Text);

            container.Should().BeEquivalentTo(new Container(Text));
        }
    }

    public class Test2
    {
        static Test2()
        {
            AssertionOptions.AssertEquivalencyUsing(
                options => options
                    .Using<Container>(
                        context =>
                        {
                            context.Subject.Should().BeEquivalentTo(context.Expectation);
                        })
                    .WhenTypeIs<Container>());
        }

        [Fact]
        public void Test()
        {
            const string Text = "Some text";
            var x = new Container(Text);

            x.Should().BeEquivalentTo(new Container(Text));
        }
    }

    public class Container
    {
        public Container(string text) => this.Text = text;

        public string Text { get; }
    }
}
```

### Expected behavior:

Have Test 1 run normally
Have Test 2 run with updated AssertionOptions

### Actual behavior:

I get this exception.
```
System.InvalidOperationException : Collection was modified; enumeration operation may not execute.
  Stack Trace:
   at System.Collections.Generic.List`1.Enumerator.MoveNextRare()
   at System.Linq.Enumerable.ConcatIterator`1.MoveNext()
   at System.Collections.Generic.List`1.InsertRange(Int32 index, IEnumerable`1 collection)
   at System.Collections.Generic.List`1.AddRange(IEnumerable`1 collection)
   at FluentAssertions.Equivalency.SelfReferenceEquivalencyAssertionOptions`1..ctor(IEquivalencyAssertionOptions defaults)
   at FluentAssertions.Equivalency.EquivalencyAssertionOptions`1..ctor(IEquivalencyAssertionOptions defaults)
   at FluentAssertions.AssertionOptions.CloneDefaults[T]()
   at FluentAssertions.Primitives.ObjectAssertions.BeEquivalentTo[TExpectation](TExpectation expectation, Func`2 config, String because, Object[] becauseArgs)
   at FluentAssertions.Primitives.ObjectAssertions.BeEquivalentTo[TExpectation](TExpectation expectation, String because, Object[] becauseArgs)
```

### Versions

* Which version of Fluent Assertions are you using?
5.10.3
* Which .NET runtime and version are you targeting? E.g. .NET framework 4.6.1 or .NET Core 2.1.
.NET Core 3.1 (netcoreapp3.1)

### Additional Information

Any additional information, configuration or data that might be necessary to reproduce the issue.

**Comment by philippdolder** [comment]
I'll try to get my hands dirty with this one in the next couple of days. Though I have absolutely no clue how on earth I would write a reliable test that verifies this issue is actually gone 🤔

**Comment by jnyrup** [comment]
Thanks for looking into this area.
I've never given much thought about having custom `AssertionOptions` for only a subset of tests.

I think the first question we need to answer before wanting to make `AssertionOptions` thread-safe is:
How is `AssertionOptions` supposed to work at all?
If it's not meant to be called concurrently, there's no need to make it thread-safe.
Note: I haven't thought about this answer, so please don't take this a rejection.

E.g. for my tests at work with MsTestV2 `AssertionOptions` works as expected (read: expected by me).
`AssertionOptions.AssertEquivalencyUsing` is only invoked a single time for the entire test suite and is shared across *all* test classes/methods.

```cs
[TestClass]
public class TestClass1
{
    [TestMethod]
    public void TestMethod1()
    {
        // uses the globally overridden AssertionOptions
    }
}

[TestClass]
public class TestClass2
{
    [TestMethod]
    public void TestMethod2()
    {
        // uses the globally overridden AssertionOptions
    }
}

[TestClass]
public static class TestInitializer
{
    [AssemblyInitialize]
    public static void InitializeTestAssembly(TestContext context)
    {
        AssertionOptions.AssertEquivalencyUsing(...);
    }
}
```

If I understand your example you'd like to use the default `AssertionOptions` for all tests in `Test1` and the custom `AssertionOptions` for all tests in `Test2`.
Correct me if I'm wrong.

**Comment by philippdolder** [comment]
> I've never given much thought about having custom `AssertionOptions` for only a subset of tests.
> If I understand your example you'd like to use the default `AssertionOptions` for all tests in `Test1` and the custom `AssertionOptions` for all tests in `Test2`.

Maybe the code sample given is a bit over-simplified. In our particular case we have some custom types that require special treatment to behave properly with `BeEquivalentTo()`.

E.g.

```
public class MyCustomType { ... content omitted ... }
public class Container { public MyCustomType Value {get;set;} }

var x = new Container {Value = new MyCustomType()};
x.Should().BeEquivalentTo(...)
```
So, some tests will require our customized AssertionOptions, some simply don't care. There are *no* tests which explicitly *don't* want the customized AssertionOptions in place. We really would want them to be globally. So since we don't have a way (that we know) to achieve that purely with xUnit.net, we tried to go with a different approach: Every test that requires them will just ensure the assertion options are there. That's why we came up with [this piece of code](https://github.com/fluentassertions/fluentassertions/issues/808#issuecomment-625054280)

And that's where it gets nasty when tests are running in parallel, which they should as much as possible.

> I think the first question we need to answer before wanting to make `AssertionOptions` thread-safe is:
> How is `AssertionOptions` supposed to work at all?
> If it's not meant to be called concurrently, there's no need to make it thread-safe.
> Note: I haven't thought about this answer, so please don't take this a rejection.

I completely agree with you on this question. There needs to be clarity about that.

> E.g. for my tests at work with MsTestV2 `AssertionOptions` works as expected (read: expected by me).

I think one of the challenges is, that every testing framework has different possibilities to hook into levels of *"global"* initialization and these sort of things.

After reading through the xUnit.net docs you referred to in #808 some more I see one approach that might work.
- Use the collection fixtures feature
- Make a collection definition in every test assembly
- Put all tests that require these assertions in the same collection.

This has some issues though. As far as I understand a test class can only be part of a single collection. We're already using test collections for some other sort of grouping (e.g. tests against the real database are in a collection). These tests also use the customized AssertionOptions so that's a problem. And the collection fixture will serialize the execution of all tests in the collection, so this might increase runtime for all tests significantly when a large amount requires the customized AssertionOptions.

Hopefully I could clarify our case a bit

**Comment by jnyrup** [comment]
Maybe this project solves your problem by adding assembly level fixtures. 
https://github.com/JDCain/Xunit.Extensions.AssemblyFixture

**Comment by philippdolder** [comment]
interesting, didn't know that. We consider using either this extension (with the drawback of having to explicitly declare the assembly fixture interface in every test. Or use fody init as Urs mentioned in #808.

Maybe these resolves this issue, and we could update the docs accordingly. If it indeed resolves our issue I'd happily contribute the docs to explain the usage of AssertionOptions further for the xUnit.net case, and you could add your bits for MsTest?

**Comment by jnyrup** [comment]
I just tried fiddling with creating a custom `XunitTestFramework` and registering it using `[assembly: TestFramework()]`.

So far my experiments show that it:
* is invoked exactly once,
* before any tests are run and
* lets tests run in parallel.

[XUnitTestProject3.zip](https://github.com/fluentassertions/fluentassertions/files/4598235/XUnitTestProject3.zip)

**Comment by philippdolder** [comment]
@jnyrup your effort and dedication is amazing. I will check that out as well.

In the meantime we figured out an adjustment to the initial `Configure()` that works (no test failures by running tests continuously for an hour on command line and inside rider)

```c#
public static class GlobalFluentAssertionsConfiguration
{
    private static readonly object Lock = new object();
    private static bool alreadyLoaded = false;

    public static void UseMaybe()
    {
        lock (Lock)
        {
            if (alreadyLoaded)
            {
                return;
            }

            alreadyLoaded = true;

            // create a bare option instead of modifying with each Using call
            var defaultWithCustom = new EquivalencyAssertionOptions()
                .Using<>().WhenTypeIs<>()
                .Using<>().WhenTypeIs<>();

            AssertionOptions.AssertEquivalencyUsing(
                options => defaultWithCustom);
        }
    }
}
```

does that make sense to you? basically replacing at once, instead of modifying in a few goes

**Comment by eNeRGy164** [comment]
You could add an additional `if (alreadyLoaded) { return; }` before the lock, if it's already true, nobody will set it to false and saves a lock

**Comment by iappert** [comment]
We're currently testing @jnyrup excellent idea, so far it runs without any problems. Good one.

**Comment by jnyrup** [comment]
I would go great lengths to avoid relying on static constructors.

```cs
class A
{
}

class B
{
    static B()
    {
        // init stuff
    }  
}
```

The static constructor [guarantee](https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/classes-and-structs/static-constructors) is as follows:
> It is called automatically before the first instance is created or any static members are referenced.

So whether `B` is referenced, and hence its static constructor is invoked, before running any test in `A` is up to the test runner. 
I certainly wouldn't rely on how/when a particular test runner references types.

I spent five minutes fiddling with this in VS with `xunit.runner.visualstudio`.
When running all tests the static constructor of `B` would always be run before any tests.
But when running only tests of `A` the static constructor was not invoked.
That makes tests in class `A` fragile if their outcome relies/changes whether the static constructor was invoked.
That's a no-go for reliable unit tests.

**Comment by philippdolder** [comment]
> I would go great lengths to avoid relying on static constructors.

Exactly, they're fragile to rely on. That's why with our first approach we called the `Configure` in the static constructor of every class that requires the custom assertions.

Btw. @iappert is in the team with me :-)

**Comment by philippdolder** [comment]
we believe we've come to a stable solution with only using built-in xunit functionality, as suggested by @jnyrup.
In every test assembly you have to add this class including the `assembly` attribute.
All of our attempts to move the fixture code to a shared assembly, and only having the `assembly` attribute in every test assembly, failed. It's also important to note that you can't use `nameof` to reference the fixture name. Xunit won't pick it up.

```c#
[assembly: Xunit.TestFramework("MyNamespace.MyFixture", "MyAssembly.Facts")]

namespace MyNamespace
{
    using Xunit.Abstractions;
    using Xunit.Sdk;

    public class MyFixture: XunitTestFramework
    {
        public MyFixture(IMessageSink messageSink)
            : base(messageSink)
        {
            GlobalFluentAssertionsConfiguration.Configure();
        }
    }
}
```

the global configuration can be shared by using an assembly reference though, as expected. We left the lock in there, but I'm quite sure we could safely remove it.
```c#
public static class GlobalFluentAssertionsConfiguration
{
    private static readonly object Lock = new object();
    private static bool alreadyLoaded = false;

    public static void UseMaybe()
    {
        if (alreadyLoaded)
        {
            return;
        }
        lock (Lock)
        {
            if (alreadyLoaded)
            {
                return;
            }

            alreadyLoaded = true;

            // create a bare option instead of modifying with each Using call
            var defaultWithCustom = new EquivalencyAssertionOptions()
                .Using<>().WhenTypeIs<>()
                .Using<>().WhenTypeIs<>();

            AssertionOptions.AssertEquivalencyUsing(
                options => defaultWithCustom);
        }
    }
}
```

If you guys @dennisdoomen @jnyrup could point me to an appropriate place in the docs, I'll create a PR to include this, without using the locking and alreadyLoaded check

**Comment by dennisdoomen** [comment]
Awesome result. I think it should end up under https://fluentassertions.com/tips/ (which go [here](https://github.com/fluentassertions/fluentassertions/tree/develop/docs/_pages)).

**Comment by philippdolder** [comment]
> I succeeded in creating a shared library and including it in two separate test projects
 
I'll check it again, maybe we overlooked something on our side.

**Comment by philippdolder** [comment]
I'm not quite sure what we did wrong in our first tries. But I got it working like a charm now. Will create a PR in the next couple of days

**Comment by philippdolder** [comment]
I think you could change the label for this issue from `bug` to `docs`?

**Comment by jnyrup** [comment]
It just hit me that [`Module Initializers`](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/proposals/csharp-9.0/module-initializers) seems to fit perfectly here.

```cs
internal static class FluentAssertionsInit
{
    [System.Runtime.CompilerServices.ModuleInitializer]
    internal static void InitGlobalOptions()
    {
        FluentAssertions.AssertionOptions.AssertEquivalencyUsing(
            options => options.WithoutStrictOrdering());
    }
}
```
